### DISCUSSION TITLE
Make file name of attachment available as key

### DISCUSSION THREAD

**Discussion by floli** [trganda/obsidian-attachment-management#40] [Q&A]
Given the list of keys from the option page, there is no option to use the original filename of the attachment, i.e. what I want to have `file.pdf` renamed to `2023-05-16 file.pdf`.

What be cool to have `{{attachmentFileName}}` available.

**Comment by trganda** [comment]
Need a variable to represent the old name of attach (before last rename operation)?
Or the original name before using this plugin?

i.e. 
origianl name: `file.pdf`
after first time renamed name: `2023-05-16 file.pdf`
after second time renamed name: `2023-05-17 2023-05-16 file.pdf` or `2023-05-17 file.pdf`

**Comment by floli** [comment]
What I need is the original file name, i.e., the name before I moved it into Obsidian.

**Comment by trganda** [comment]
This feature need to persisten the attachment name, with the growing of vault, the cost of this plugin will be heavy.

Btw, we could add a option to enable append the attachment name (set in settings) to prefix or suffix of the attachment that need to be renamed.

**Comment by floli** [comment]
My intention was not keep the original name for later renamings, but only use for the initial renaming when a file is added to the vault. Later, of course, the original file is lost.

Best!

**Comment by trganda** [comment]
I have know you request.

We could add a variable like ${attachmentFileName} in `Attachment format` field, and if the variable was used for one note, the auto rename will disable for it.

Currently, If you don't need to renaming the file in future but only for first adding to vault, your can turn off the `Automatically rename attachment`, of cause it will close the auto rename globaly right now.

**Comment by Mara-Li** [comment]
It is maybe possible to keep the original filename on a variable, the first time the file is added to a vault, and during the renaming, in place of using the date (for example), apply the original filename? Or maybe rename it "two times" ?  
I think you can apply the new path & apply the original name right after ?

**Comment by trganda** [comment]
Hi, @floli.
You can use ${originalname} as the original name in attachment name in [0.5.0](https://github.com/trganda/obsidian-attachment-management/releases/tag/0.5.0).

Tks for @Lisandra-dev's work.

**Comment by breakstar** [comment]
![image](https://github.com/trganda/obsidian-attachment-management/assets/29723651/8ba77ab4-265a-4224-b31d-47087cb5879d)

I'm sorry to bother you.
Your software is very practical, but I encountered a little problem in the process of use, I would like to consult you.
I used {originalname}}, copied the image in Explorer, then pasted it twice in OB editor, there was always no renames, and the two photos were stored in different paths.

**Comment by trganda** [comment]
> ![image](https://user-images.githubusercontent.com/29723651/255856747-8ba77ab4-265a-4224-b31d-47087cb5879d.png)
> 
> I'm sorry to bother you.很抱歉打扰你。 Your software is very practical, but I encountered a little problem in the process of use, I would like to consult you.你们的软件很实用，但是在使用过程中遇到了一点问题，想咨询一下。 I used {originalname}}, copied the image in Explorer, then pasted it twice in OB editor, there was always no renames, and the two photos were stored in different paths.我使用了{originalname}}，在资源管理器中复制了图像，然后在OB编辑器中粘贴了两次，始终没有重命名，并且两张照片存储在不同的路径中。

The ${originalname} will be replace with the filename that you pasted.
And what's your filename? `Snipaste_2023-07-25_11-51-41 1.jpg`?

Have you modifed the **Attachment format** between twice paste?

I have try to set **Attachment format** as `IMG-${originalname}` and paste a png with name `Crayon.png` in same file. The result is below

<img width="361" alt="image" src="https://github.com/trganda/obsidian-attachment-management/assets/62204882/9d12f684-1f13-4581-954f-aa0bff45cfe7">

Can you explain more detail of your situation

**Comment by breakstar** [comment]
> > ![image](https://user-images.githubusercontent.com/29723651/255856747-8ba77ab4-265a-4224-b31d-47087cb5879d.png)
> > I'm sorry to bother you.很抱歉打扰你。 Your software is very practical, but I encountered a little problem in the process of use, I would like to consult you.你们的软件很实用，但是在使用过程中遇到了一点问题，想咨询一下。 I used {originalname}}, copied the image in Explorer, then pasted it twice in OB editor, there was always no renames, and the two photos were stored in different paths.我使用了{originalname}}，在资源管理器中复制了图像，然后在OB编辑器中粘贴了两次，始终没有重命名，并且两张照片存储在不同的路径中。
> 
> The ${originalname} will be replace with the filename that you pasted. And what's your filename? `Snipaste_2023-07-25_11-51-41 1.jpg`?
> 
> Have you modifed the **Attachment format** between twice paste?
> 
> I have try to set **Attachment format** as `IMG-${originalname}` and paste a png with name `Crayon.png` in same file. The result is below
> 
> <img alt="image" width="361" src="https://user-images.githubusercontent.com/62204882/255860080-9d12f684-1f13-4581-954f-aa0bff45cfe7.png">
> Can you explain more detail of your situation





Thank you very much for your reply.
The source file name is: 'Snipaste_2023-07-25_11-51-41.jpg', copy and paste from the Explorer to the ob editor, do not make any changes, continue to paste to the ob editor, there is no rename situation, and the paths of the two images are different.
attachment-format: IMG-${originalname}-${date}

**Reply by trganda** [reply]
I have tried your setting, but it's not happen to me. 

Looks like your first paste was use a setting like `${originalname}`, and second `IMG-${originalname}-${date}`. I don't know if the filename is 'Snipaste_2023-07-25_11-51-41.jpg' where is the suffix ` 1` and ` 2` from. For duplicate file the suffix will be `-{number}`.

Please show me the version number of this plugin and paste the content of `.obsidian/plugins/attachment-management/data.json`.

> 另外，你可以用中文。

**Reply by breakstar** [reply]
好的。我详细描述下。

原文件：
![image](https://github.com/trganda/obsidian-attachment-management/assets/29723651/772b8456-be1a-4918-83ee-fad7c697860e)

对该文件在资源管理器复制后，在ob编辑器连续粘贴4次结果如图：
> 粘贴结果与第2次，第4次与预期不符。且第二次存储路径是安装系统设置的attachment目录，而不是插件设置的目录。

![image](https://github.com/trganda/obsidian-attachment-management/assets/29723651/1c8a856c-9aa2-46cc-a589-31e029b22baf)


`.obsidian/plugins/attachment-management/data.json.` 如下：
```
{
  "attachPath": {
    "attachmentRoot": "Attachment",
    "saveAttE": "inFolderBelow",
    "attachmentPath": "${notepath}/${notename}",
    "attachFormat": "IMG-${originalname}-${date}",
    "type": "GLOBAL"
  },
  "dateFormat": "YYYYMMDDHHmmssSSS",
  "handleAll": false,
  "excludeExtensionPattern": "",
  "autoRenameAttachment": true,
  "excludedPaths": "",
  "excludePathsArray": [],
  "excludeSubpaths": false,
  "overridePath": {}
}
````

插件信息如下：
![image](https://github.com/trganda/obsidian-attachment-management/assets/29723651/ef254df2-2eab-4a89-9352-1b5025e860ec)

**Reply by trganda** [reply]
你好，感谢你反馈的信息。

首先我在 Mac 和 Windows 下进行了测试，但是只有 Windows 中出现了你描述的问题。首先导致这个问题产生的原因之一是下面的这段代码
见 https://github.com/trganda/obsidian-attachment-management/blob/main/src/main.ts#L141C1-L144C12。

```
const timeGapMs = new Date().getTime() - file.stat.ctime;
if (timeGapMs > 1000) {
  return;
}
```

这段代码意思是，对于ob中创建的文件，如果文件的创建时间和当前时间之间的时间间隔大于 1 秒，那么就忽略这个文件不进行处理。

我不太清楚是什么原因导致在 windows 中会有这样的现象发生（大概是不同操作系统之间的差异），我尝试关闭所有插件，重复粘贴同一文件观察它们的创建时间

<img width="421" alt="image" src="https://github.com/trganda/obsidian-attachment-management/assets/62204882/fd4be1ca-2d5e-4c42-9f75-97ca003be0ad">

下面是原文件的创建时间

<img width="434" alt="image" src="https://github.com/trganda/obsidian-attachment-management/assets/62204882/8de6c624-c44a-4d23-a9f3-7d62e4bb2ab8">

但进行重复粘贴操作的时间大约是 `11:21`。

稍后我会进行修复，只需要将 `file.stat.ctime` 改为 `file.stat.mtime`，但需要注意的是，如果你以非常快的速度（大约小于1s的时间间隔）进行连续粘贴的操作，可能还是会出现前面的问题.

**Reply by trganda** [reply]
你可以尝试更新到最新版本了

https://github.com/trganda/obsidian-attachment-management/releases/tag/0.7.3

**Reply by breakstar** [reply]
好的，感谢你的修复。稍后我会安装新版本进行尝试。
